### PR TITLE
Extending Tessera to Support Private-slash-Point-to-Point Messaging

### PR DESCRIPTION
**Pull request by viathefalcon**
Using Tessera for (asynchronous) messaging in this way is structurally similar to how it is used for propagating private transactions: a client seeking to send a message to some other entity makes a request to an endpoint on its Tessera node, specifying the message along with the public key of the recipient. The Tessera Enclave would generate a (shared) key between the private key corresponding to the sender's public key and the recipient's public key and use the shared key to encrypt the message contents which would be (synchronously) forwarded on to the recipient's Tessera node from which the recipient can retrieve and decrypt the message, on demand.

Key differences with how private transactions are handled by Tessera include:
- Messages are only persisted on the receiving side
- Messages are not intended to be long-lived; clients are expected to delete messages once they've been read/processed
- Barring some other mechanism for notifying clients that a new message has arrived, requiring their attention, clients would need to poll for new messages (and to delete read messages to make polling less inefficient)

### CODE REVIEW

File: tessera-core/src/main/java/com/quorum/tessera/messaging/internal/InboxImpl.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    } catch (Exception ex) {
+      LOGGER.warn("Failed to instantiate a required message digest", ex);
+      throw new IllegalStateException(ex);
+    }
+  }
+
+  MessageHash hash(byte[] encoded) {
+    sha256.reset();
+    sha1.reset();
+    return new MessageHash(sha1.digest(sha256.digest(encoded)));
```
[the remark this answers is not in the document]
Comment (viathefalcon) [reply tessera-core/src/main/java/com/quorum/tessera/messaging/internal/InboxImpl.java:38]: There weren't any requirements as such (certainly not security); we've previously used a similar technique for generating content-based identifiers in our own blockchain/blockchain-adjacent projects, so I simply re-applied it here.

Restricting it to just SHA-1, if required/desirable, wouldn't be a problem.

File: tessera-jaxrs/sync-jaxrs/src/main/java/com/quorum/tessera/p2p/MessageResource.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                  @Schema(
+                      description = "identifier of received message",
+                      type = "string",
+                      format = "base64")))
+  @PUT
+  @Path("/push")
+  @Consumes(APPLICATION_OCTET_STREAM)
+  public Response push(@Schema(description = "arbitrary message") final byte[] message) {
+
+    LOGGER.info(
```
[the remark this answers is not in the document]
Comment (viathefalcon) [reply tessera-jaxrs/sync-jaxrs/src/main/java/com/quorum/tessera/p2p/MessageResource.java]: Yes, good point; fixed.

File: tessera-data/src/main/java/com/quorum/tessera/data/EncryptedMessage.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  public void setTo(byte[] to) {
+    this.to = to;
+  }
+  */
+
+  @Override
+  public int hashCode() {
+    return 47 * 3 + Objects.hashCode(this.hash);
```
Comment (taccatisid) [inline tessera-data/src/main/java/com/quorum/tessera/data/EncryptedMessage.java:110]: This mistakes operator precedence, effectively just adding 127. The multiplicative constant should also be a larger prime.

Comment (viathefalcon) [reply tessera-data/src/main/java/com/quorum/tessera/data/EncryptedMessage.java:110]: In that case, https://github.com/ConsenSys/tessera/blob/8f340564ea0a97b82560226ab6d2b72e07b1b519/tessera-data/src/main/java/com/quorum/tessera/data/EncryptedTransaction.java#L75 is equally deficient (it's where I copied this from).

But, if you can suggest a better implementation, I'd be happy to adopt it, thanks.

Comment (taccatisid) [reply tessera-data/src/main/java/com/quorum/tessera/data/EncryptedMessage.java:110]: `return 174407 + Objects.hashCode(this.hash);`

**Pull request closed without merging**